File: python/packages/botas/src/botas/clients/bot_http_client.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, TypeVar
from urllib.parse import urlencode

import httpx

TokenProvider = Callable[[], Awaitable[str | None]]

T = TypeVar("T")


@dataclass
class BotRequestOptions:
    operation_description: str = ""
    return_none_on_not_found: bool = False


class BotHttpError(Exception):
    def __init__(self, message: str, status_code: int, body: str = "") -> None:
        super().__init__(message)
        self.status_code = status_code
        self.body = body


class BotHttpClient:
    def __init__(self, get_token: TokenProvider | None = None) -> None:
        self._get_token = get_token
        self._client = httpx.AsyncClient(
            headers={"Content-Type": "application/json"},
            timeout=30.0,
        )

    async def _auth_headers(self) -> dict[str, str]:
        if self._get_token is None:
            return {}
        token = await self._get_token()
        if token is None:
            return {}
        return {"Authorization": f"Bearer {token}"}

    def _build_url(
        self, base_url: str, endpoint: str, params: dict[str, str | None] | None
    ) -> str:
        url = base_url.rstrip("/") + endpoint
        if params:
            filtered = {k: v for k, v in params.items() if v is not None}
            if filtered:
                url += "?" + urlencode(filtered)
        return url

    async def _send(
        self,
        method: str,
        url: str,
        body: Any,
        options: BotRequestOptions,
    ) -> Any:
        headers = await self._auth_headers()
        resp = await self._client.request(
            method,
            url,
            json=body,
            headers=headers,
        )
        if resp.status_code == 404 and options.return_none_on_not_found:
            return None
        if resp.status_code == 204:
            return None
        if not resp.is_success:
            raise BotHttpError(
                f"{options.operation_description} failed: HTTP {resp.status_code}",
                resp.status_code,
                resp.text,
            )
        if not resp.content:
            return None
        return resp.json()

    async def get(
        self,
        base_url: str,
        endpoint: str,
        params: dict[str, str | None] | None = None,
        options: BotRequestOptions | None = None,
    ) -> Any:
        opts = options or BotRequestOptions()
        url = self._build_url(base_url, endpoint, params)
        return await self._send("GET", url, None, opts)

    async def post(
        self,
        base_url: str,
        endpoint: str,
        body: Any,
        options: BotRequestOptions | None = None,
    ) -> Any:
        opts = options or BotRequestOptions()
        url = self._build_url(base_url, endpoint, None)
        return await self._send("POST", url, body, opts)

    async def put(
        self,
        base_url: str,
        endpoint: str,
        body: Any,
        options: BotRequestOptions | None = None,
    ) -> Any:
        opts = options or BotRequestOptions()
        url = self._build_url(base_url, endpoint, None)
        return await self._send("PUT", url, body, opts)

    async def delete(
        self,
        base_url: str,
        endpoint: str,
        options: BotRequestOptions | None = None,
    ) -> None:
        opts = options or BotRequestOptions()
        url = self._build_url(base_url, endpoint, None)
        await self._send("DELETE", url, None, opts)

    async def aclose(self) -> None:
        await self._client.aclose()

File: node/packages/botas-core/src/bot-http-client.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import axios, { type AxiosInstance, type AxiosRequestConfig } from 'axios'
import { getLogger } from './logger.js'

/** Options for a single HTTP request made by {@link BotHttpClient}. */
export interface BotRequestOptions {
  /** Human-readable description of the operation (used in error messages). */
  operationDescription?: string;
  /** When `true`, a 404 response returns `undefined` instead of throwing. */
  returnNullOnNotFound?: boolean;
  /** Additional request headers to merge in. */
  headers?: Record<string, string>;
}

/** A function that returns a Bearer token for the outgoing request. */
export type TokenProvider = () => Promise<string>

/**
 * Authenticated HTTP client for the Bot Framework REST API.
 *
 * Wraps axios and injects a Bearer token via a request interceptor when a
 * {@link TokenProvider} is supplied.
 *
 * **Lifecycle note:** Create one `BotHttpClient` instance per bot application
 * and reuse it. The request interceptor is attached once in the constructor.
 * Creating many short-lived instances is safe but wasteful — interceptors do
 * not accumulate within a single instance.
 *
 * **Response validation:** Callers are responsible for validating response
 * shapes. The generic type parameter `T` provides compile-time convenience
 * but no runtime schema enforcement.
 */
export class BotHttpClient {
  private readonly http: AxiosInstance

  /**
   * Create a new authenticated HTTP client.
   *
   * @param getToken - Optional token provider. When supplied, every outgoing
   *   request receives a `Bearer` token via an Axios request interceptor.
   *   Omit for unauthenticated calls (dev/testing only).
   */
  constructor (getToken?: TokenProvider) {
    this.http = axios.create({
      headers: { 'Content-Type': 'application/json' },
      timeout: 30_000,
    })

    if (getToken) {
      this.http.interceptors.request.use(async (config) => {
        const token = await getToken()
        config.headers.Authorization = `Bearer ${token}`
        return config
      })
    }
  }

  /**
   * Send an HTTP request and return the parsed response body.
   *
   * @typeParam T - Expected shape of the response body (not validated at runtime).
   * @param method - HTTP method (e.g. `"GET"`, `"POST"`).
   * @param url - Fully-qualified URL to request.
   * @param body - Optional request body, serialized as JSON.
   * @param options - Additional request options (error handling, headers).
   * @returns The parsed response body, or `undefined` if the response has no content
   *   or if `returnNullOnNotFound` is set and a 404 was received.
   * @throws {Error} On non-404 HTTP errors, with the method, URL, and status in the message.
   */
  async send<T>(
    method: string,
    url: string,
    body?: unknown,
    options: BotRequestOptions = {}
  ): Promise<T | undefined> {
    const config: AxiosRequestConfig = {
      method,
      url,
      data: body ?? undefined,
      headers: options.headers,
    }

    try {
      const response = await this.http.request<T>(config)
      return response.data ?? undefined
    } catch (err: unknown) {
      if (isAxiosError(err)) {
        if (options.returnNullOnNotFound && err.response?.status === 404) {
          return undefined
        }
        const status = err.response?.status ?? 'no response'
        // #93: Log full upstream body at debug level only; never leak in error messages
        if (err.response?.data !== undefined) {
          getLogger().debug('%s %s upstream response body: %s', method, url, JSON.stringify(err.response.data))
        }
        throw new Error(`${method} ${url} failed with status ${status}`)
      }
      throw err
    }
  }

  /**
   * Send a GET request to a Bot Framework API endpoint.
   *
   * @typeParam T - Expected response body type.
   * @param baseUrl - Service URL base (e.g. `activity.serviceUrl`).
   * @param endpoint - API path (e.g. `/v3/conversations`).
   * @param params - Optional query string parameters; `undefined` values are omitted.
   * @param options - Additional request options.
   * @returns Parsed response body, or `undefined`.
   */
  async get<T>(
    baseUrl: string,
    endpoint: string,
    params?: Record<string, string | undefined>,
    options?: BotRequestOptions
  ): Promise<T | undefined> {
    const url = buildUrl(baseUrl, endpoint, params)
    return this.send<T>('GET', url, undefined, options)
  }

  /**
   * Send a POST request to a Bot Framework API endpoint.
   *
   * @typeParam T - Expected response body type.
   * @param baseUrl - Service URL base.
   * @param endpoint - API path.
   * @param body - Request body, serialized as JSON.
   * @param options - Additional request options.
   * @param params - Optional query string parameters.
   * @returns Parsed response body, or `undefined`.
   */
  async post<T>(
    baseUrl: string,
    endpoint: string,
    body?: unknown,
    options?: BotRequestOptions,
    params?: Record<string, string | undefined>
  ): Promise<T | undefined> {
    const url = buildUrl(baseUrl, endpoint, params)
    return this.send<T>('POST', url, body, options)
  }

  /**
   * Send a PUT request to a Bot Framework API endpoint.
   *
   * @typeParam T - Expected response body type.
   * @param baseUrl - Service URL base.
   * @param endpoint - API path.
   * @param body - Request body, serialized as JSON.
   * @param options - Additional request options.
   * @param params - Optional query string parameters.
   * @returns Parsed response body, or `undefined`.
   */
  async put<T>(
    baseUrl: string,
    endpoint: string,
    body?: unknown,
    options?: BotRequestOptions,
    params?: Record<string, string | undefined>
  ): Promise<T | undefined> {
    const url = buildUrl(baseUrl, endpoint, params)
    return this.send<T>('PUT', url, body, options)
  }

  /**
   * Send a DELETE request to a Bot Framework API endpoint.
   *
   * @param baseUrl - Service URL base.
   * @param endpoint - API path.
   * @param params - Optional query string parameters.
   * @param options - Additional request options.
   */
  async delete (
    baseUrl: string,
    endpoint: string,
    params?: Record<string, string | undefined>,
    options?: BotRequestOptions
  ): Promise<void> {
    const url = buildUrl(baseUrl, endpoint, params)
    await this.send<void>('DELETE', url, undefined, options)
  }
}

function buildUrl (
  baseUrl: string,
  endpoint: string,
  params?: Record<string, string | undefined>
): string {
  const base = baseUrl.replace(/\/$/, '')
  const path = endpoint.startsWith('/') ? endpoint : `/${endpoint}`
  let url = `${base}${path}`

  if (params) {
    const query = Object.entries(params)
      .filter(([, v]) => v !== undefined)
      .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v!)}`)
      .join('&')
    if (query) url += `?${query}`
  }

  return url
}

function isAxiosError (err: unknown): err is { response?: { status: number; data?: unknown } } {
  return typeof err === 'object' && err !== null && 'response' in err
}
